### PR TITLE
Upgrade AO aux tables using relfilenodes

### PR DESCRIPTION
**Pull request by kyeap-vmware**
The branch is based on bmdoil's pg_dump_6x_perf which has not been merged yet.
**We just need to review the last 4 commits.**

Previously, AO aux tables (pg_aoseg, pg_aovisimap, pg_aoblkdir) were
upgraded by querying the old database, then rebuilt on the new
cluster using TRUNCATES and INSERTS. This was likely done because
pg_aoseg catalog changes from GPDB4 to GPDB5 make relfilenodes between
these two version incompatible. Pg_aoseg tables have remain unchanged
since GPDB5. We can upgrade them using their relfilenodes. This is done
by leveraging the existing logic that already copies other relation's
relfilenodes.

GPDB5 pg_aoseg catalog change reference commit:
https://github.com/greenplum-db/gpdb/commit/53590b354b5

gpupgrade pipeline: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:6X_ao_relfilenode
6X pipeline: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/6X_RC_ao_relfilenode
6X pipeline with icw tests: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-pg_upgrade-aotables-using-relfilenode 

Some of these queries can probably be optimized. 

Next steps things to be done.
- Check if commit https://github.com/greenplum-db/gpdb/commit/8be6d344175 is still valid. This issue could be fixed by the _fsm _vm file removals on segments.
- get green 6X pipeline. Only 1 red job before my last rebase failure related to commit above.
- get green gpupgrade pipeline. it is currently failing because pg_dump diffs.
- Write additional tests for edge cases, will likely land in 
https://github.com/greenplum-db/gpupgrade/tree/main/test/acceptance/pg_upgrade.
- get perf numbers on local machine with varying number of AO tables.
- get perf numbers by running against real world schemas.

### CODE REVIEW

File: contrib/pg_upgrade/info.c [hunk 1]
```diff
@@ -67,6 +71,19 @@ gen_db_file_maps(DbInfo *old_db, DbInfo *new_db,
...
+		 * */
+		if ((strcmp(old_rel->nspname, "pg_aoseg") == 0) &&
+			(strncmp(old_rel->relname, "pg_aoblkdir", 11) == 0))
+		{
+			old_relnum++;
+			/* new_relnum++; */
+			continue;
+		}
+
+
```
Comment (jimmyyih) [inline contrib/pg_upgrade/info.c:86]: The old patch was more correct.  This one filters out all pg_aoblkdir which we've learned is not correct since the pg_aoblkdir for existing AO indexes need to be preserved through upgrade.  That being said, there were some minor issues with the old patch's queries as well.

```
    PQclear(executeQueryOrDie(conn,
                "DELETE FROM info_rels "
                "WHERE reloid IN ("
                "   SELECT cb.oid "
                "   FROM pg_appendonly a "
                "   LEFT JOIN pg_class cb ON (cb.oid = a.blkdirrelid "
                "       AND NOT EXISTS ("
                "           SELECT 1 "
                "           FROM pg_index i "
                "           WHERE i.indrelid = a.relid)))"));
    PQclear(executeQueryOrDie(conn,
                "DELETE FROM info_rels "
                "WHERE reloid IN ("
                "   SELECT cb.oid "
                "   FROM pg_appendonly a "
                "   LEFT JOIN pg_class cb ON (cb.oid = a.blkdiridxid "
                "       AND NOT EXISTS ("
                "           SELECT 1 "
                "           FROM pg_index i "
                "           WHERE i.indrelid = a.relid)))"));
```

1. The two queries could be combined by just using `cb.oid IN (a.blkdirrelid, a.blkdiridxid)`.
2. The LEFT JOIN causes NULL results for each table with a valid AO index.  Changing it to just a regular JOIN fixes that issue.

However, I recommend using this query instead which is seemingly more performant (after comparing EXPLAIN ANALYZE query plans):
```
DELETE FROM info_rels WHERE reloid IN (
    SELECT c.oid
    FROM pg_class c
    JOIN pg_appendonly a ON c.oid IN (a.blkdirrelid, a.blkdiridxid)
    LEFT JOIN pg_index i ON i.indrelid = a.relid
    WHERE i.indexrelid IS NULL);
```

File: src/bin/pg_dump/pg_dump.c [hunk 2]
```diff
@@ -14613,6 +14708,29 @@ dumpIndex(Archive *fout, IndxInfo *indxinfo)
...
+					tbinfo->aotbl->blkdirfrozenxid, tbinfo->aotbl->blkdirminmxid, tbinfo->aotbl->blkdirrelid);
+
+			appendPQExpBufferStr(q, "\n-- For binary upgrade, set child AO partition's pg_aoblkdir relfrozenxid and relminmxid\n");
+			appendPQExpBuffer(q, "UPDATE pg_catalog.pg_class\n"
+					"SET relfrozenxid = '%u', relminmxid = '%u'\n"
+					"WHERE oid IN (select c.oid from pg_inherits i JOIN pg_appendonly ao on ao.relid=i.inhrelid JOIN pg_class c on c.oid=ao.blkdirrelid where inhparent='%u');\n",
+					tbinfo->aotbl->blkdirfrozenxid, tbinfo->aotbl->blkdirminmxid, tbinfo->dobj.catId.oid);
+
+			appendPQExpBuffer(q, "RESET allow_system_table_mods;\n");
+		}
```
Comment (jimmyyih) [inline src/bin/pg_dump/pg_dump.c:14732]: This gets dumped multiple times if the AO table has multiple indexes... but I guess that's alright for now since I can't think of a quick clean way to fix that (another reason why aoblkdir creation on CREATE INDEX instead of CREATE TABLE can be annoying for upgrade).

File: src/bin/pg_dump/pg_dump.c [hunk 3]
```diff
@@ -14189,6 +14223,67 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
...
+				partres = ExecuteSqlQuery(fout, partquery->data, PGRES_TUPLES_OK);
+
+				if (PQntuples(partres) > 0)
+				{
+					int			i;
+
+					for (i = 0; i < PQntuples(partres); i++)
+					{
+						Oid part_oid = atooid(PQgetvalue(partres, i, 0));
+						TableInfo *tbinfo = findTableByOid(part_oid);
```
Comment (jimmyyih) [inline src/bin/pg_dump/pg_dump.c:14265]: While valid, we really shouldn't define a local `tbinfo` variable as it can cause confusion with the outer-defined `tbinfo` function arg.  Please rename it to something like `part_tbinfo` and change the callers accordingly.

File: src/bin/pg_dump/pg_dump.c [hunk 4]
```diff
@@ -14613,6 +14708,29 @@ dumpIndex(Archive *fout, IndxInfo *indxinfo)
...
+			appendPQExpBuffer(q, "UPDATE pg_catalog.pg_class\n"
+					"SET relfrozenxid = '%u', relminmxid = '%u'\n"
+					"WHERE oid = '%u'::pg_catalog.oid;\n",
+					tbinfo->aotbl->blkdirfrozenxid, tbinfo->aotbl->blkdirminmxid, tbinfo->aotbl->blkdirrelid);
+
+			appendPQExpBufferStr(q, "\n-- For binary upgrade, set child AO partition's pg_aoblkdir relfrozenxid and relminmxid\n");
+			appendPQExpBuffer(q, "UPDATE pg_catalog.pg_class\n"
+					"SET relfrozenxid = '%u', relminmxid = '%u'\n"
+					"WHERE oid IN (select c.oid from pg_inherits i JOIN pg_appendonly ao on ao.relid=i.inhrelid JOIN pg_class c on c.oid=ao.blkdirrelid where inhparent='%u');\n",
+					tbinfo->aotbl->blkdirfrozenxid, tbinfo->aotbl->blkdirminmxid, tbinfo->dobj.catId.oid);
```
Comment (jimmyyih) [inline src/bin/pg_dump/pg_dump.c:14729]: Am I correct in thinking that the `relfrozenxid` is being set to the same value (the root partition's value) for every partition's aoblkdir?  That doesn't seem right.

File: contrib/pg_upgrade/info.c [hunk 5]
```diff
@@ -551,6 +536,26 @@ get_rel_infos(ClusterInfo *cluster, DbInfo *dbinfo)
 							  "     ON ind.indrelid = i.reloid "
 							  "WHERE indisvalid AND i.toastheap != %u", InvalidOid));
 
+	/*
+	 * Resolve the edge case where an aoblkdir and it's index exists, but the AO
+	 * table it was created for no longer has an index. The aoblkdir is created
+	 * the first time an index is placed on an AO table. If the last index on
+	 * the AO table is deleted, the aoblkdir is not removed even though it is
+	 * unused. This is a known behavior.
+	 * https://github.com/greenplum-db/gpdb/pull/13887.
```
Comment (jimmyyih) [inline contrib/pg_upgrade/info.c:545]: Should put the full link to the discussion you're referencing:
https://github.com/greenplum-db/gpdb/pull/13887#discussion_r938966260

Or not give the link at all.  The overall comment describes the issue relatively well already.

File: src/bin/pg_dump/pg_dump.c (hunk 4 above)
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c:14729]: Yes you're right. It can be fixed with some minor refactoring. Relevant comment:
```
		 * GPDB: To gather binary upgrade information for partition children,
		 * they need to exist in the tblinfo array for findTableByOid.
		 * This requires the getTable query to also collect all
		 * partition children so they can be referenced, but we do not want
		 * to dump the partition children as their DDL will be handled by the parent.
		 * We do need to dump external partition DDL however, so ensure partition children
		 * with external storage have their dump flag set.
```
We have the child partition's relfrozenxid for aoseg/blkdir/visimap so need to utilize it correctly. The `UPDATE` can't be set in `dumpIndex` because only the parent index is set to dumpable. 

A few months ago I was working on a refactor to the partition parent specific block of `dumpTableSchema` - dropping the extra `SELECT DISTINCT` queries for in favor of setting up a partition inheritance structure by parsing a single query result, then referencing the cached part children results when handling the parent in `dumpTableSchema`. I didn't get to finish the work but could dig it up for reference.

File: src/bin/pg_dump/pg_dump.c (hunk 2 above)
Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c:14732]: We can move this out of dumpIndex. The check `if (tbinfo->aotbl && (OidIsValid(tbinfo->aotbl->blkdirrelid)))` should be in `dumpTableSchema` because it's not relevant to the current state of indexes on the table.

File: contrib/pg_upgrade/relfilenode.c [hunk 6]
```diff
@@ -287,6 +303,29 @@ transfer_relfile_segment(int segno, FileNameMap *map, const char *type_suffix)
...
+	 * upgrade segments, coordinator's relfilenodes, _fsm, _vm files will end up
+	 * on segments. Normally this is ok since the files will end up being
+	 * overwritten. However, there is an edge case where there can be data in a
+	 * table on coordinator, but no data on the segment. Examples of such tables
+	 * where this can occur are pg_ao(cs)seg tables. If this edge case happens,
+	 * The new segment's table will end up with old segment's relfilenode and new
+	 * coordinator's _fsm and _vm file. The _fsm and _vm files don't get
+	 * overwritten because they aren't supposed to exist on the segment.
+	 * Attempting to run VACUUM on this table after upgrade completes will
+	 * result in a similar error below.
```
Comment (bmdoil) [inline contrib/pg_upgrade/relfilenode.c:316]: Has there been any consideration to adding an exclude list when copying MDD to segments, so we don't end up with unwanted objects in the first place? Also saves some CPU cycles.

**Comment by kyeap-vmware** [comment]
@jimmyyih @bmdoil 

After thinking about it bit more and discussing it with Jimmy, we realized that pg_dump changes should not be needed anymore.

Originally, pg_dump changes were needed because segment upgrade would fail due to pg_aoseg relfrozenxid issues from the coordinator. These are resolved from gp_fastsequence commit where coordinator now also gets relfrozenxid fixes that were previously only for segments. Reference commit https://github.com/greenplum-db/gpdb/commit/9cd8cd2630.

The failure on segments were the result of 2 things.
1. Coordinator was frozen before relfilenodes were copied from source to target cluster. Which means tuples that were not frozen may come in after the coordinator was frozen.
2. During coordinator upgrade, pg_aoseg relations did not get their refrozenxids modified to ensure xmin < relfrozenxid.

Both of these issues were fixed in the gp_fastsequence commit in order to move gp_fastsequence values to the target cluster using relfilenodes. I have reverted the pg_dump changes and pushed them to the branch, the 6X_STABLE and gpupgrade pipelines are still passing.

File: contrib/pg_upgrade/relfilenode.c (hunk 6 above)
Comment (kyeap-vmware) [reply contrib/pg_upgrade/relfilenode.c:316]: No consideration for this PR. If we do consider an exclusion list will be part of another PR.

File: contrib/pg_upgrade/pg_upgrade.c [hunk 7]
```diff
@@ -215,16 +215,6 @@ main(int argc, char **argv)
...
 
-	/*
-	 * In a segment, the data directory already contains all the objects,
-	 * because the segment is initialized by taking a physical copy of the
-	 * upgraded QD data directory. The auxiliary AO tables - containing
-	 * information about the segment files, are different in each server,
-	 * however. So we still need to restore those separately on each
-	 * server.
-	 */
-	restore_aosegment_tables();
```
[the remark this answers is not in the document]
Comment (kyeap-vmware) [reply contrib/pg_upgrade/pg_upgrade.c:226]: Removed restore_aosegment_tables and is_appendonly function declaration.

File: contrib/pg_upgrade/info.c [hunk 8]
```diff
@@ -551,6 +536,25 @@ get_rel_infos(ClusterInfo *cluster, DbInfo *dbinfo)
 							  "     ON ind.indrelid = i.reloid "
 							  "WHERE indisvalid AND i.toastheap != %u", InvalidOid));
 
+	/*
+	 * Resolve the edge case where an aoblkdir and it's index exists, but the AO
+	 * table it was created for no longer has any indexes. The aoblkdir is
+	 * created the first time an index is placed on an AO table. If all indexes
+	 * on for the table are dropped, the aoblkdir is not removed even though it
```
Comment (jimmyyih) [inline contrib/pg_upgrade/info.c:543]: s/on for/on/

File: contrib/pg_upgrade/info.c [hunk 9]
```diff
@@ -551,6 +536,25 @@ get_rel_infos(ClusterInfo *cluster, DbInfo *dbinfo)
...
+	 * is unused. This is a known behavior. The edge case will result in an
+	 * upgrade failure when relations between the old and new clusters are
+	 * compared. The aoblkdirs and its index would exist on the old cluster, but
```
Comment (jimmyyih) [inline contrib/pg_upgrade/info.c:546]: s/its index/their indexes/

**Review by jimmyyih — APPROVED** [review]
Looks good after fixing the minor comment stuff.
